File: python/compresr/credentials.py
"""Persistent credentials store at ``~/.compresr/credentials``."""

from __future__ import annotations

import configparser
import os
import re
import stat
import sys
import tempfile
import time
from contextlib import contextmanager
from pathlib import Path
from typing import Iterator, Optional

DEFAULT_PROFILE = "default"

_REPLACE_RETRIES = 6
_REPLACE_BACKOFF_S = 0.1
_LOCK_DEADLINE_S = 10.0
_LOCK_POLL_S = 0.1

_API_KEY_RE = re.compile(r"^cmp_[A-Za-z0-9_-]{16,128}$")


class LockAcquisitionError(RuntimeError):
    """Raised when the credentials lock can't be acquired in time."""


def credentials_path() -> Path:
    override = os.environ.get("COMPRESR_CREDENTIALS_FILE")
    if override:
        return Path(override).expanduser()
    return Path.home() / ".compresr" / "credentials"


def _new_parser() -> configparser.ConfigParser:
    return configparser.ConfigParser(interpolation=None)


def _refuse_symlink(path: Path, kind: str) -> None:
    if path.is_symlink():
        raise PermissionError(f"Refusing to use {kind} {path}: symlink at that path.")


def _read_file() -> configparser.ConfigParser:
    parser = _new_parser()
    path = credentials_path()
    if not path.exists():
        return parser
    _refuse_symlink(path, "credentials file")
    if os.name == "posix":
        mode = path.lstat().st_mode
        if mode & (stat.S_IRWXG | stat.S_IRWXO):
            raise PermissionError(
                f"Refusing to read {path}: file is group/world accessible. "
                "Run `chmod 600` on it."
            )
    parser.read(path)
    return parser


@contextmanager
def _file_lock(path: Path) -> Iterator[None]:
    path.parent.mkdir(parents=True, exist_ok=True)
    lock_path = path.with_suffix(path.suffix + ".lock")

    if os.name == "posix":
        import fcntl

        fd = os.open(str(lock_path), os.O_RDWR | os.O_CREAT, 0o600)
        acquired = False
        try:
            deadline = time.monotonic() + _LOCK_DEADLINE_S
            while True:
                try:
                    fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
                    acquired = True
                    break
                except (BlockingIOError, OSError):
                    if time.monotonic() >= deadline:
                        raise LockAcquisitionError(
                            f"Could not acquire {lock_path} within "
                            f"{_LOCK_DEADLINE_S:g}s — another process may be "
                            "writing, or a previous run left the lock stale."
                        )
                    time.sleep(_LOCK_POLL_S)
            yield
        finally:
            if acquired:
                try:
                    fcntl.flock(fd, fcntl.LOCK_UN)
                except OSError:
                    pass
            os.close(fd)
        return

    try:
        import msvcrt  # type: ignore[import-not-found]
    except ImportError:
        print(
            "warning: file locking unavailable on this platform; "
            "concurrent `compresr-sdk` invocations may race.",
            file=sys.stderr,
        )
        yield
        return
    fd = os.open(str(lock_path), os.O_RDWR | os.O_CREAT)
    acquired = False
    try:
        deadline = time.monotonic() + _LOCK_DEADLINE_S
        while True:
            try:
                msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)  # type: ignore[attr-defined]
                acquired = True
                break
            except OSError:
                if time.monotonic() >= deadline:
                    raise LockAcquisitionError(
                        f"Could not acquire {lock_path} within {_LOCK_DEADLINE_S:g}s."
                    )
                time.sleep(_LOCK_POLL_S)
        yield
    finally:
        if acquired:
            try:
                msvcrt.locking(fd, msvcrt.LK_UNLCK, 1)  # type: ignore[attr-defined]
            except OSError:
                pass
        os.close(fd)


def _atomic_write(parser: configparser.ConfigParser, path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if os.name == "posix":
        default_dir = Path.home() / ".compresr"
        if path.parent == default_dir:
            try:
                os.chmod(path.parent, 0o700)
            except OSError:
                pass

    tmp_path: Optional[Path] = None
    replaced = False
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", dir=str(path.parent), delete=False, prefix=".credentials-"
        ) as tmp:
            tmp_path = Path(tmp.name)
            parser.write(tmp)
            tmp.flush()
            os.fsync(tmp.fileno())

        if os.name == "posix":
            os.chmod(tmp_path, 0o600)

        if path.is_symlink():
            raise PermissionError(f"Refusing to replace symlink at {path}; remove it first.")

        last_err: Optional[OSError] = None
        for attempt in range(_REPLACE_RETRIES):
            try:
                os.replace(tmp_path, path)
                replaced = True
                break
            except OSError as e:
                last_err = e
                time.sleep(_REPLACE_BACKOFF_S * (2**attempt))
        if not replaced and last_err is not None:
            raise last_err
    finally:
        if not replaced and tmp_path is not None and tmp_path.exists():
            try:
                tmp_path.unlink()
            except OSError:
                pass


def _safe_read_or_fresh(path: Path) -> configparser.ConfigParser:
    try:
        return _read_file()
    except configparser.Error as e:
        print(
            f"warning: credentials file at {path} is corrupt ({e}); recreating.",
            file=sys.stderr,
        )
        return _new_parser()


def load(profile: str = DEFAULT_PROFILE) -> Optional[str]:
    """Return the stored API key for ``profile``, or ``None``."""
    try:
        parser = _read_file()
    except configparser.Error:
        return None
    if profile not in parser:
        return None
    return parser[profile].get("api_key") or None


def save(
    api_key: str,
    *,
    profile: str = DEFAULT_PROFILE,
    base_url: Optional[str] = None,
    account_type: Optional[str] = None,
) -> Path:
    from datetime import datetime, timezone

    if not _API_KEY_RE.match(api_key):
        raise ValueError(
            "Refusing to save malformed api key: " "must match ^cmp_[A-Za-z0-9_-]{16,128}$."
        )

    path = credentials_path()
    with _file_lock(path):
        parser = _safe_read_or_fresh(path) if path.exists() else _new_parser()
        section = {
            "api_key": api_key,
            "saved_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        }
        if base_url:
            section["base_url"] = base_url
        if account_type:
            section["account_type"] = account_type
        parser[profile] = section
        _atomic_write(parser, path)
    return path


def clear(profile: str = DEFAULT_PROFILE) -> bool:
    path = credentials_path()
    if not path.exists():
        return False
    with _file_lock(path):
        try:
            parser = _read_file()
        except configparser.Error:
            return False
        if profile not in parser:
            return False
        parser.remove_section(profile)
        _atomic_write(parser, path)
    return True


def resolve_api_key(explicit: Optional[str], profile: Optional[str] = None) -> Optional[str]:
    """explicit > env > credentials file. An explicit non-None short-circuits
    fallbacks (even ``""``) so downstream validation can distinguish "user
    passed empty" from "no arg passed"."""
    if explicit is not None:
        return explicit
    env = os.environ.get("COMPRESR_API_KEY")
    if env:
        return env
    try:
        return load(profile or os.environ.get("COMPRESR_PROFILE", DEFAULT_PROFILE))
    except PermissionError:
        return None


__all__ = [
    "credentials_path",
    "load",
    "save",
    "clear",
    "resolve_api_key",
    "DEFAULT_PROFILE",
    "LockAcquisitionError",
]

File: typescript/src/auth/credentials.ts
/**
 * Persistent credentials store at ``~/.compresr/credentials.json``. Node-only.
 */

import { randomBytes, randomUUID } from 'node:crypto';
import {
  chmodSync,
  closeSync,
  constants as fsConstants,
  existsSync,
  fsyncSync,
  lstatSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  unlinkSync,
  writeSync,
} from 'node:fs';
import { hostname, homedir } from 'node:os';
import { dirname, join } from 'node:path';

export const DEFAULT_PROFILE = 'default';

const REPLACE_RETRIES = 6;
const REPLACE_BACKOFF_MS = 100;
const LOCK_ATTEMPTS = 60;
const LOCK_BACKOFF_MS = 100;
const LOCK_STALE_MS = 30_000;

const S_IFLNK = 0o120000;
const S_IFMT = 0o170000;
const S_IRWXG = 0o070;
const S_IRWXO = 0o007;

const POSIX_TMP_FLAGS =
  fsConstants.O_WRONLY |
  fsConstants.O_CREAT |
  fsConstants.O_EXCL |
  (fsConstants.O_NOFOLLOW ?? 0);

export const API_KEY_RE = /^cmp_[A-Za-z0-9_-]{16,128}$/;

export interface CredentialsSection {
  api_key: string;
  saved_at: string;
  base_url?: string;
  account_type?: string;
}

interface CredentialsFile {
  [profile: string]: CredentialsSection;
}

export class BadPermissionsError extends Error {
  constructor(public readonly path: string) {
    super(
      `Refusing to read ${path}: file is group/world accessible. ` +
        'Run `chmod 600` on it.'
    );
    this.name = 'BadPermissionsError';
  }
}

export class LockAcquisitionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'LockAcquisitionError';
  }
}

export function credentialsPath(): string {
  const override = process.env.COMPRESR_CREDENTIALS_FILE;
  if (override) {
    return override.startsWith('~')
      ? join(homedir(), override.slice(1))
      : override;
  }
  return join(homedir(), '.compresr', 'credentials.json');
}

function isPosix(): boolean {
  return process.platform !== 'win32';
}

function isSymlink(path: string): boolean {
  try {
    return (lstatSync(path).mode & S_IFMT) === S_IFLNK;
  } catch {
    return false;
  }
}

function assertNotSymlink(path: string, kind: string): void {
  if (isSymlink(path)) {
    throw new BadPermissionsError(
      `Refusing to use ${kind} at ${path}: symlink not permitted.`
    );
  }
}

function assertSafePerms(path: string): void {
  if (!isPosix()) return;
  if (lstatSync(path).mode & (S_IRWXG | S_IRWXO)) {
    throw new BadPermissionsError(path);
  }
}

function isSection(v: unknown): v is CredentialsSection {
  if (typeof v !== 'object' || v === null) return false;
  const o = v as Record<string, unknown>;
  return typeof o.api_key === 'string' && typeof o.saved_at === 'string';
}

function readFile(): CredentialsFile {
  const path = credentialsPath();
  if (!existsSync(path)) return {};
  assertNotSymlink(path, 'credentials file');
  assertSafePerms(path);
  const raw = readFileSync(path, 'utf-8');
  if (raw.trim() === '') return {};
  const parsed: unknown = JSON.parse(raw);
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw new SyntaxError('credentials file is not a JSON object');
  }
  for (const [profile, section] of Object.entries(parsed as Record<string, unknown>)) {
    if (!isSection(section)) {
      throw new SyntaxError(
        `credentials profile '${profile}' is malformed (missing api_key or saved_at)`
      );
    }
  }
  return parsed as CredentialsFile;
}

function safeReadOrFresh(): CredentialsFile {
  try {
    return readFile();
  } catch (e) {
    if (e instanceof BadPermissionsError) throw e;
    if (e instanceof SyntaxError) {
      process.stderr.write(
        `warning: credentials file at ${credentialsPath()} is corrupt ` +
          `(${e.message}); recreating.\n`
      );
      return {};
    }
    throw e;
  }
}

function sleep(ms: number): void {
  try {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
  } catch {
    const end = Date.now() + Math.min(ms, 50);
    while (Date.now() < end) {
      /* spin */
    }
  }
}

interface LockOwner {
  pid: number;
  uuid: string;
  hostname: string;
  ts: number;
}

function isLockOwnerLive(owner: LockOwner): boolean {
  if (owner.hostname !== hostname()) return true;
  try {
    process.kill(owner.pid, 0);
    return true;
  } catch {
    return false;
  }
}

function readLockOwner(lockPath: string): LockOwner | null {
  try {
    const parsed: unknown = JSON.parse(readFileSync(lockPath, 'utf-8'));
    if (
      typeof parsed !== 'object' ||
      parsed === null ||
      typeof (parsed as LockOwner).pid !== 'number' ||
      typeof (parsed as LockOwner).uuid !== 'string' ||
      typeof (parsed as LockOwner).hostname !== 'string'
    ) return null;
    return parsed as LockOwner;
  } catch {
    return null;
  }
}

function tryStealLock(lockPath: string, existingUuid: string): boolean {
  // CAS on the uuid — protects against a race with the original owner
  // refreshing the lock between our checks.
  const current = readLockOwner(lockPath);
  if (current?.uuid !== existingUuid) return false;
  try {
    unlinkSync(lockPath);
    return true;
  } catch {
    return false;
  }
}

function acquireLock<T>(path: string, work: () => T): T {
  const lockPath = `${path}.lock`;
  mkdirSync(dirname(lockPath), { recursive: true });

  const owner: LockOwner = {
    pid: process.pid,
    uuid: randomUUID(),
    hostname: hostname(),
    ts: Date.now(),
  };
  const ownerJson = JSON.stringify(owner);

  let held = false;
  for (let attempt = 0; attempt < LOCK_ATTEMPTS; attempt++) {
    try {
      const fd = openSync(lockPath, 'wx', 0o600);
      try {
        writeSync(fd, ownerJson);
      } finally {
        closeSync(fd);
      }
      held = true;
      break;
    } catch (e) {
      const err = e as NodeJS.ErrnoException;
      if (err.code !== 'EEXIST') throw e;
      const existing = readLockOwner(lockPath);
      const age = existing ? Date.now() - existing.ts : Number.MAX_SAFE_INTEGER;
      if (existing && !isLockOwnerLive(existing) && age > 1_000) {
        tryStealLock(lockPath, existing.uuid);
        continue;
      }
      if (existing && age > LOCK_STALE_MS) {
        tryStealLock(lockPath, existing.uuid);
        continue;
      }
      sleep(LOCK_BACKOFF_MS);
    }
  }

  if (!held) {
    throw new LockAcquisitionError(
      `Could not acquire ${lockPath} within ` +
        `${((LOCK_ATTEMPTS * LOCK_BACKOFF_MS) / 1000).toFixed(1)}s — ` +
        `another process may be writing, or a previous run left the lock stale.`
    );
  }

  try {
    return work();
  } finally {
    try {
      const current = readLockOwner(lockPath);
      if (!current || current.uuid === owner.uuid) {
        unlinkSync(lockPath);
      }
    } catch {
      /* ignore */
    }
  }
}

function fsyncDirectory(dir: string): void {
  if (!isPosix()) return;
  try {
    const fd = openSync(dir, 0);
    try {
      fsyncSync(fd);
    } finally {
      closeSync(fd);
    }
  } catch {
    /* ENOTSUP on some FSes */
  }
}

function atomicWrite(data: CredentialsFile, path: string): void {
  const dir = dirname(path);
  mkdirSync(dir, { recursive: true });
  const defaultDir = join(homedir(), '.compresr');
  if (isPosix() && dir === defaultDir) {
    try { chmodSync(dir, 0o700); } catch { /* not fatal */ }
  }

  const tmpPath = join(
    dir,
    `.credentials-${process.pid}-${randomBytes(8).toString('hex')}`
  );
  const openFlag = isPosix() ? POSIX_TMP_FLAGS : 'wx';
  let written = false;
  let replaced = false;

  try {
    const fd = openSync(tmpPath, openFlag, 0o600);
    try {
      writeSync(fd, JSON.stringify(data, null, 2) + '\n');
      fsyncSync(fd);
      written = true;
    } finally {
      closeSync(fd);
    }
    if (isPosix()) {
      try { chmodSync(tmpPath, 0o600); } catch { /* not fatal */ }
    }

    if (isSymlink(path)) {
      throw new BadPermissionsError(
        `Refusing to replace symlink at ${path}; remove it first.`
      );
    }

    let lastErr: unknown = null;
    for (let attempt = 0; attempt < REPLACE_RETRIES; attempt++) {
      try {
        renameSync(tmpPath, path);
        replaced = true;
        break;
      } catch (e) {
        lastErr = e;
        sleep(REPLACE_BACKOFF_MS * 2 ** attempt);
      }
    }
    if (!replaced) throw lastErr;

    fsyncDirectory(dir);
  } finally {
    if (written && !replaced) {
      try { unlinkSync(tmpPath); } catch { /* ignore */ }
    }
  }
}

export function load(profile: string = DEFAULT_PROFILE): string | null {
  let data: CredentialsFile;
  try {
    data = readFile();
  } catch (e) {
    if (e instanceof BadPermissionsError) throw e;
    return null;
  }
  return data[profile]?.api_key || null;
}

export interface SaveOptions {
  profile?: string;
  baseUrl?: string;
  accountType?: string;
}

export function save(apiKey: string, opts: SaveOptions = {}): string {
  if (!API_KEY_RE.test(apiKey)) {
    throw new Error(
      'Refusing to save malformed api key: must match ^cmp_[A-Za-z0-9_-]{16,128}$.'
    );
  }
  const profile = opts.profile ?? DEFAULT_PROFILE;
  const path = credentialsPath();
  acquireLock(path, () => {
    const data = existsSync(path) ? safeReadOrFresh() : {};
    const section: CredentialsSection = {
      api_key: apiKey,
      saved_at: new Date().toISOString(),
    };
    if (opts.baseUrl) section.base_url = opts.baseUrl;
    if (opts.accountType) section.account_type = opts.accountType;
    data[profile] = section;
    atomicWrite(data, path);
  });
  return path;
}

export function clear(profile: string = DEFAULT_PROFILE): boolean {
  const path = credentialsPath();
  if (!existsSync(path)) return false;
  let removed = false;
  acquireLock(path, () => {
    let data: CredentialsFile;
    try {
      data = readFile();
    } catch (e) {
      if (e instanceof BadPermissionsError) throw e;
      return;
    }
    if (!(profile in data)) return;
    delete data[profile];
    atomicWrite(data, path);
    removed = true;
  });
  return removed;
}

export function listProfiles(): string[] {
  try {
    return Object.keys(readFile());
  } catch (e) {
    if (e instanceof BadPermissionsError) throw e;
    return [];
  }
}
